client/layout/guided-tours/config-elements/step.tsx

```diff
@@ -96,43 +94,46 @@ export default class Step extends Component< Props, State > {
 	 */
 	isUpdatingPosition: boolean = false;
 
-	componentWillMount() {
-		this.wait( this.props, this.context ).then( () => {
-			this.start();
-			this.setStepSection( this.context, { init: true } );
-			debug( 'Step#componentWillMount: stepSection:', this.stepSection );
-			this.skipIfInvalidContext( this.props, this.context );
-			this.scrollContainer = query( this.props.scrollContainer )[ 0 ] || window;
-			this.setStepPosition( this.props );
-			this.safeSetState( { initialized: true } );
-		} );
-	}
+	async componentDidMount() {
+		await this.wait( this.props, this.context );
+
+		this.start();
+		this.setStepSection( this.context, { init: true } );
+		debug( 'Step#componentDidMount: stepSection:', this.stepSection );
+		this.skipIfInvalidContext( this.props, this.context );
+		this.scrollContainer = query( this.props.scrollContainer )[ 0 ] || window;
+		this.setStepPosition( this.props );
+		// The following should be legit usage of `setState`, since it's called
+		// asynchronously after `await`ing `this.wait()`.
+		// (We need it since `this.render()` might be called in the meantime.)
+		// eslint-disable-next-line react/no-did-mount-set-state
+		this.setState( { initialized: true } );
+		window.addEventListener( 'resize', this.onScrollOrResize );
+		this.watchTarget();
 
-	componentDidMount() {
-		this.mounted = true;
-		this.wait( this.props, this.context ).then( () => {
-			window.addEventListener( 'resize', this.onScrollOrResize );
-			this.watchTarget();
-		} );
 		if ( this.props.keepRepositioning ) {
 			this.repositionInterval = setInterval( this.onScrollOrResize, 3000 );
 		}
 	}
 
-	componentWillReceiveProps( nextProps: Props, nextContext ) {
+	async componentWillReceiveProps( nextProps: Props, nextContext ) {
 		const shouldScrollTo = nextProps.shouldScrollTo && this.props.name !== nextProps.name;
-		this.wait( nextProps, nextContext ).then( () => {
-			this.setStepSection( nextContext );
-			this.quitIfInvalidRoute( nextProps, nextContext );
-			this.skipIfInvalidContext( nextProps, nextContext );
-			if ( this.scrollContainer ) {
-				this.scrollContainer.removeEventListener( 'scroll', this.onScrollOrResize );
-			}
-			this.scrollContainer = query( nextProps.scrollContainer )[ 0 ] || window;
-			this.scrollContainer.addEventListener( 'scroll', this.onScrollOrResize );
-			this.setStepPosition( nextProps, shouldScrollTo );
-			this.watchTarget();
-		} );
+
+		if ( nextProps.wait !== this.props.wait ) {
+			await this.wait( nextProps, nextContext );
+		}
```

Comment:
This doesn't seem quite right. What if the promise is unresolved but wait hasn't changed? The awaiting `componentDidMount` will not have resumed, but we'll run `componentWillReceiveProps` without waiting at all.